### PR TITLE
`sage.{modular,schemes}`: Modularization fixes for imports; update `sage -fiximports`, add relint pattern 

### PR DESCRIPTION
**Pull request by mkoeppe**
### :books: Description

- Part of: #29705 
- Cherry-picked from: #35095 

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [ ] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### :hourglass: Dependencies

- Depends on #35974 (to soothe the linter)

### CODE REVIEW

**Comment by mkoeppe** [comment]
The Linter failure is unrelated and fixed in #35974

File: src/sage/schemes/elliptic_curves/padic_lseries.py [hunk 1]
```diff
@@ -61,7 +61,7 @@
 #                  https://www.gnu.org/licenses/
 ######################################################################
 
-import sage.matrix.all as matrix
+from sage.matrix.constructor import Matrix as matrix
 import sage.schemes.hyperelliptic_curves.monsky_washnitzer
```
Comment (kwankyu) [inline src/sage/schemes/elliptic_curves/padic_lseries.py:65]: why not just `from sage.matrix.constructor import matrix`?

File: src/sage/schemes/elliptic_curves/ell_tate_curve.py [hunk 2]
```diff
@@ -50,7 +50,7 @@
 from sage.functions.log import log
 from sage.misc.functional import denominator
 from sage.misc.misc_c import prod
-import sage.matrix.all as matrix
+from sage.matrix.constructor import Matrix as matrix
```
Comment (kwankyu) [inline src/sage/schemes/elliptic_curves/ell_tate_curve.py:54]: why not `from sage.matrix.constructor import matrix`?

File: src/sage/modular/pollack_stevens/padic_lseries.py [hunk 3]
```diff
@@ -21,13 +21,15 @@
...
+from sage.misc.lazy_import import lazy_import
 from sage.rings.integer_ring import ZZ
-from sage.rings.padics.factory import Qp as pAdicField
 from sage.rings.padics.precision_error import PrecisionError
 from sage.rings.power_series_ring import PowerSeriesRing
 from sage.rings.rational_field import QQ
 from sage.structure.sage_object import SageObject
 
+lazy_import('sage.rings.padics.factory', 'Qp', as_='pAdicField')
+
```
Comment (kwankyu) [inline src/sage/modular/pollack_stevens/padic_lseries.py:32]: Why only this is lazy-imported while `PrecisionError` from `sage.rings.padics.precision_error` is just imported?

File: src/sage/modular/modsym/space.py [hunk 4]
```diff
@@ -46,7 +47,7 @@
 from sage.modular.hecke.module import HeckeModule_free_module
 from sage.modular.modsym.element import ModularSymbolsElement
 
-from . import hecke_operator
+lazy_import('sage.modular.modsym', 'hecke_operator')
```
Comment (kwankyu) [inline src/sage/modular/modsym/space.py:51]: Why is this lazy-imported?

File: src/sage/modular/modsym/all.py [hunk 5]
```diff
@@ -1,9 +1,10 @@
+from sage.misc.lazy_import import lazy_import
 
 from .element import set_modsym_print_mode
 
 from .modsym import ModularSymbols, ModularSymbols_clear_cache
 
-from .heilbronn import HeilbronnCremona, HeilbronnMerel
+lazy_import('sage.modular.modsym.heilbronn', ['HeilbronnCremona', 'HeilbronnMerel'])
```
Comment (kwankyu) [inline src/sage/modular/modsym/all.py:8]: How do you decide that this should be lazy-imported?

File: src/sage/modular/pollack_stevens/padic_lseries.py (hunk 3 above)
Comment (mkoeppe) [reply src/sage/modular/pollack_stevens/padic_lseries.py:32]: sage.rings.padics.precision_error is shipped by **sagemath-categories**

One cannot do try: ... except ... with a LazyImport

File: src/sage/modular/modsym/all.py (hunk 5 above)
Comment (mkoeppe) [reply src/sage/modular/modsym/all.py:8]: It has a compile-time dependency on FLINT and won't be shipped in sagemath-schemes

File: src/sage/modular/modsym/space.py (hunk 4 above)
Comment (mkoeppe) [reply src/sage/modular/modsym/space.py:51]: IIRC, via .hecke it depends on .arithgroup, which I had some difficulties with

File: src/sage/schemes/elliptic_curves/padic_lseries.py (hunk 1 above)
Comment (mkoeppe) [reply src/sage/schemes/elliptic_curves/padic_lseries.py:65]: OK, done (in all places)

File: src/.relint.yml [hunk 6]
```diff
@@ -53,6 +53,17 @@
   # imports from .all are allowed in all.py; also allow in some modules that need sage.all
   filePattern: '(.*/|)(?!(all|benchmark|dev_tools|parsing|sage_eval))[^/.]*[.](py|pyx|pxi)$'
 
+- name: 'namespace_pkg_all_import_2: Module-level import of .all of a namespace package'
+  hint: |
+    Sage library code should not import sage.PAC.KAGE.all when sage.PAC.KAGE is an implicit
+    namespace package. Type import_statements("SOME_IDENTIFIER") to find a more specific import,
+    and rewrite the import statement as "from sage.PAC.KAGE.MODULE import ..."
+    or "lazy_import('sage.PAC.KAGE.MODULE', '...')".
+  # Keep in sync with above; but for now we ignore sage.{arith,categories}
```
Comment (kwankyu) [inline src/.relint.yml:62]: You may want to edit the content of `hint:`.

File: src/.relint.yml [hunk 7]
```diff
@@ -53,6 +53,17 @@
...
+  pattern: '^import\s+sage(|[.](combinat|crypto|databases|data_structures|dynamics|ext|game_theory|games|geometry|graphs|groups|interfaces|manifolds|matrix|matroids|misc|modules|monoids|numerical|probability|quadratic_forms|quivers|rings|sat|schemes|sets|stats|symbolic|tensor)[a-z0-9_.]*|[.]libs)[.]all'
+  # imports from .all are allowed in all.py; also allow in some modules that need sage.all
+  filePattern: '(.*/|)(?!(all|benchmark|dev_tools|parsing|sage_eval|explain_pickle|.*_test))[^/.]*[.](py|pyx|pxi)$'
+
```
Comment (kwankyu) [inline src/.relint.yml:66]: You don't add `explain_pickle` also in L54?

File: src/sage/misc/replace_dot_all.py [hunk 8]
```diff
@@ -385,7 +385,11 @@ def make_replacements_in_file(location, package_regex=None, verbose=False, outpu
         write_file.write(replaced_content)  # overwriting the old file contents with the new/replaced content
 
 
-def walkdir_replace_dot_all(dir, file_regex=r'.*[.](py|pyx|pxi)$', package_regex=None, verbose=False):
+excluded_file_regex = 'auto-methods|replace_dot_all'
+
```
Comment (kwankyu) [inline src/sage/misc/replace_dot_all.py]: You may want `r'` like below

Comment (kwankyu) [reply src/sage/misc/replace_dot_all.py]: Do we actually need L388? The argument `excluded_file_regex` of `walkdir_replace_dot_all` gets the default value. No?

**Comment by kwankyu** [comment]
To justify fixes of modularization-anti-pattern imports, I think each file should be annotated with the distribution which the file belongs to (is installed by). Do we already have a standard style for the annotation? 

For `file.py`, the annotated distribution would be used as `xxx` in `sage -t --distribution=xxx file.py`. Right?

I think that the future incremental test workflow should run  `sage -t --distribution=xxx file.py` for each modified `file.py`. Right?

**Comment by kwankyu** [comment]
As more files of the sage library get included to newly born distributions, the work practice of a sage developer should change much, albeit there has been little change in the practice up to now. A developer would modify a file such that the file does not break in the distribution it belongs to. If doctests that need other distributions are added, then he needs to test the doctests in another sage install where the additional distributions are available as well.

If the above is right, perhaps we should also update the developer manual to reflect the changing work practice.

In the worst, a developer would feel like that he/she is expelled from their paradise (where one can import anything from anywhere), after they were scared by the huge flood of optional tags to doctests.

File: src/.relint.yml (hunk 7 above)
Comment (mkoeppe) [reply src/.relint.yml:66]: OK, done in 5a6d6bd2a0

File: src/.relint.yml (hunk 6 above)
Comment (mkoeppe) [reply src/.relint.yml:62]: What's your suggestion here?

File: src/sage/misc/replace_dot_all.py (hunk 8 above)
Comment (mkoeppe) [reply src/sage/misc/replace_dot_all.py]: Thanks. I've removed it in 299ae874a6.

**Comment by mkoeppe** [comment]
> I think each file should be annotated with the distribution which the file belongs to (is installed by). Do we already have a standard style for the annotation?

Yes, that's `# sage_setup: distribution = sagemath-bliss`, which is currently only used for the few modules depending on optional packages. #35663 is adding many more such directives.

**Comment by mkoeppe** [comment]
> For `file.py`, the annotated distribution would be used as `xxx` in `sage -t --distribution=xxx file.py`. Right?

Yes, that's for sure one of the tests to run, but sometimes one also may want to test with extras added to enable more tests.

> I think that the future incremental test workflow should run `sage -t --distribution=xxx file.py` for each modified `file.py`.

Right.

**Comment by mkoeppe** [comment]
> As more files of the sage library get included to newly born distributions, the work practice of a sage developer should change much, albeit there has been little change in the practice up to now. A developer would modify a file such that the file does not break in the distribution it belongs to. If doctests that need other distributions are added, then he needs to test the doctests in another sage install where the additional distributions are available as well.
> 
> If the above is right, perhaps we should also update the developer manual to reflect the changing work practice.

Yes, that's a great point. While #35749 is adding a detailed description of how to test in a modularized setting, something has yet to be added to the tutorials that describe the development workflow.

> In the worst, a developer would feel like that he/she is expelled from their paradise (where one can import anything from anywhere), after they were scared by the huge flood of optional tags to doctests.

There is certainly a potential for developer frustration after we make failures of the modularized distributions hard errors. This is something to watch out for and mitigate.

**Comment by kwankyu** [comment]
> > I think each file should be annotated with the distribution which the file belongs to (is installed by). Do we already have a standard style for the annotation?
> 
> Yes, that's `# sage_setup: distribution = sagemath-bliss`, which is currently only used for the few modules depending on optional packages. #35663 is adding many more such directives.

Certainly adding those annotations is a huge task and error-prone, and it is good that there is a one source of truth. That one source of truth should be `MANIFEST.in`.

Hence it would be nice that we can add the annotations automatically from the manifest files, perhaps as part of the bootstrap. I asked chatGPT how to do that. It answered
```
python setup.py sdist --manifest-only
```
But this does not work with me. Its second answer is
```
python setup.py check --metadata --restructuredtext --strict && python setup.py sdist --formats=zip,gztar
```
This seems to work. I ran the command in `pkgs/sagemath-objects`, and got the file `pkgs/sagemath-objects/sagemath_objects.egg-info/SOURCES.txt`, which list the included files. We may use this file to add the annotations automatically.

Does this make sense?

File: src/.relint.yml (hunk 6 above)
Comment (kwankyu) [reply src/.relint.yml:62]: Ah, sorry. The hint is already perfect.

**Comment by mkoeppe** [comment]
> > that's `# sage_setup: distribution = sagemath-bliss`, which is currently only used for the few modules depending on optional packages. #35663 is adding many more such directives.
> 
> Certainly adding those annotations is a huge task and error-prone, and it is good that there is a one source of truth. That one source of truth should be `MANIFEST.in`.

Right now, this is true to some extent, but it's a bit more complicated actually. Some of the `setup.py` files also do source filtering using the `# sage_setup: distribution` directives, in addition to the filtering that MANIFEST does.

I'm hoping to unify and simplify this soon; not sure if I really want to keep the MANIFEST mechanism. 
Using the `# sage_setup: distribution` directives as the one source of truth would have the benefit that distributions are guaranteed to be disjoint.

> Hence it would be nice that we can add the annotations automatically from the manifest files, perhaps as part of the bootstrap.

This can't be part of the bootstrap because it would need to modify source files that are under version control.

But it would be indeed be very useful as a developer tool when making changes to the contents of the distributions. I'm currently doing this with very primitive tools.

**Comment by mkoeppe** [comment]
> I ran the command in `pkgs/sagemath-objects`, and got the file `pkgs/sagemath-objects/sagemath_objects.egg-info/SOURCES.txt`, which list the included files. We may use this file to add the annotations automatically.

A better command is `./sage -sh -c '(cd pkgs/sagemath-objects && python3 setup.py egg_info)'`

**Comment by kwankyu** [comment]
> > Certainly adding those annotations is a huge task and error-prone, and it is good that there is a one source of truth. That one source of truth should be `MANIFEST.in`.
> 
> Right now, this is true to some extent, but it's a bit more complicated actually. Some of the `setup.py` files also do source filtering using the `# sage_setup: distribution` directives, in addition to the filtering that MANIFEST does.
> 
> I'm hoping to unify and simplify this soon; not sure if I really want to keep the MANIFEST mechanism. Using the `# sage_setup: distribution` directives as the one source of truth would have the benefit that distributions are guaranteed to be disjoint.

We may use MANIFEST (or something renamed) as a tool to mass-add the file annotations, but use the file annotations as the one source of truth.

From developer's point of view, I think both mechanisms would be useful.

File: src/sage/modular/hecke/ambient_module.py [hunk 9]
```diff
@@ -21,9 +21,8 @@
 from . import module
 from . import submodule
 
-import sage.modules.all
-
-import sage.rings.all
+from sage.modules.free_module import FreeModule, is_FreeModule
+from sage.rings.integer import Integer
```
Comment (kwankyu) [inline src/sage/modular/hecke/ambient_module.py:26]: Couldn't we get rid of `is_FreeModule` on the way?

Comment (kwankyu) [reply src/sage/modular/hecke/ambient_module.py:26]: Perhaps not in this PR...

File: src/sage/modular/modsym/all.py (hunk 5 above)
Comment (kwankyu) [reply src/sage/modular/modsym/all.py:8]: You meant sagemath-modular?

File: src/sage/modular/modsym/space.py (hunk 4 above)
Comment (kwankyu) [reply src/sage/modular/modsym/space.py:51]: OK.

**Comment by kwankyu** [comment]
Please soothe the linter. Otherwise, looks good to me.

File: src/sage/modular/hecke/ambient_module.py (hunk 9 above)
Comment (mkoeppe) [reply src/sage/modular/hecke/ambient_module.py:26]: Right, separate PR

**Review by kwankyu — APPROVED** [review]
Thanks.

**Comment by mkoeppe** [comment]
Thank you!

**Comment by vbraun** [comment]
```
**********************************************************************
File "src/sage/schemes/elliptic_curves/hom.py", line 1163, in sage.schemes.elliptic_curves.hom.compute_trace_generic
Failed example:
    compute_trace_generic(-m7)
Exception raised:
    Traceback (most recent call last):
      File "/home/release/Sage/src/sage/doctest/forker.py", line 709, in _run
        self.compile_and_execute(example, compiler, test.globs)
      File "/home/release/Sage/src/sage/doctest/forker.py", line 1144, in compile_and_execute
        exec(compiled, globs)
      File "<doctest sage.schemes.elliptic_curves.hom.compute_trace_generic[24]>", line 1, in <module>
        compute_trace_generic(-m7)
      File "/home/release/Sage/src/sage/schemes/elliptic_curves/hom.py", line 1191, in compute_trace_generic
        P = point_of_order(E, l)
            ^^^^^^^^^^^^^^^^^^^^
      File "/home/release/Sage/src/sage/schemes/elliptic_curves/ell_field.py", line 2083, in point_of_order
        l = rings.ZZ(l)
            ^^^^^
    NameError: name 'rings' is not defined
**********************************************************************
2 items had failures:
   2 of  10 in sage.schemes.elliptic_curves.hom.EllipticCurveHom.characteristic_polynomial
   5 of  26 in sage.schemes.elliptic_curves.hom.compute_trace_generic
    [282 tests, 7 failures, 1.27 s]
----------------------------------------------------------------------
sage -t --long --warn-long 35.9 --random-seed=123 src/sage/schemes/elliptic_curves/hom.py  # 7 doctests failed
----------------------------------------------------------------------
```

**Pull request merged**